### DISCUSSION TITLE
Do typevars have inference from specification at a later point?

### DISCUSSION THREAD

**Discussion by abadrinath15** [microsoft/pyright#6919] [Q&A] [answered]
Apologies if the question isn't super clear, but I hope this example will help:

```python
import typing
P = typing.ParamSpec('P')
R = typing.TypeVar('R')
def f_with_map(map: typing.Callable[[], typing.MutableMapping[typing.Any, R]] = dict) -> typing.Callable[[typing.Callable[P, R]], typing.Callable[P, R]]:
    def inner(g: typing.Callable[P, R]) -> typing.Callable[P, R]:
        return g

    return inner

def f_without_map() -> typing.Callable[[typing.Callable[P, R]], typing.Callable[P, R]]:
    def inner(g: typing.Callable[P, R]) -> typing.Callable[P, R]:
        return g

    return inner


def sum(x: int, y: int) -> int:
    return x + y

reveal_type(f_with_map()(sum))
reveal_type(f_without_map()(sum))
```

`reveal_type(f_without_map()(sum))` looks right to me, I see it as `Type of "f_without_map()(sum)" is "(x: int, y: int) -> int"`

`reveal_type(f_with_map()(sum))` on the other hand kinda stumbles, it shows up  `"f_with_map()(sum)" is "(x: int, y: int) -> Unknown"`

I imagine this is because of the default argument of `dict`; I imagine pyright sees that as `MutableMapping[Any, Any]`, and that create a conflict with the specification coming from `sum`. What is the correct way to handle this? Should I do something like ` = dict[typing.Any, R]` in the function argument?

**Comment by erictraut** [comment, marked as answer]
Yes, the type of `dict()` is `dict[Unknown, Unknown]` which is a subtype of `MutableMapping[Any, Unknown]`, which means that type variable `R` is solved as `Unknown`.

Depending on your objectives, an overload might meet your needs.

```python
@overload
def f_with_map() -> Callable[[Callable[P, R]], Callable[P, R]]:
    ...

@overload
def f_with_map(
    map: Callable[[], MutableMapping[Any, R]]
) -> Callable[[Callable[P, R]], Callable[P, R]]:
    ...

def f_with_map(
    map: Callable[[], MutableMapping[Any, R]] = dict
) -> Callable[[Callable[P, R]], Callable[P, R]]:
    def inner(g: Callable[P, R]) -> Callable[P, R]:
        return g

    return inner
```

**Reply by abadrinath15** [reply]
Thank you Eric!
